src/runtime/virtcontainers/qemu_test.go

```diff
@@ -76,40 +76,112 @@ func TestQemuKernelParameters(t *testing.T) {
 }
 
 func TestQemuCreateVM(t *testing.T) {
-	qemuConfig := newQemuConfig()
 	assert := assert.New(t)
 
 	store, err := persist.GetDriver()
 	assert.NoError(err)
-	q := &qemu{
-		config: HypervisorConfig{
-			VMStorePath:  store.RunVMStoragePath(),
-			RunStorePath: store.RunStoragePath(),
-		},
-	}
-	sandbox := &Sandbox{
-		ctx: context.Background(),
-		id:  "testSandbox",
-		config: &SandboxConfig{
-			HypervisorConfig: qemuConfig,
-		},
-	}
 
 	// Create the hypervisor fake binary
 	testQemuPath := filepath.Join(testDir, testHypervisor)
 	_, err = os.Create(testQemuPath)
 	assert.NoError(err)
 
 	// Create parent dir path for hypervisor.json
-	parentDir := filepath.Join(store.RunStoragePath(), sandbox.id)
+	parentDir := filepath.Join(store.RunStoragePath(), "testSandbox")
 	assert.NoError(os.MkdirAll(parentDir, DirMode))
 
 	network, err := NewNetwork()
 	assert.NoError(err)
-	err = q.CreateVM(context.Background(), sandbox.id, network, &sandbox.config.HypervisorConfig)
-	assert.NoError(err)
+
+	type testData struct {
+		config      HypervisorConfig
+		expectError bool
+		configMatch bool
+	}
+
+	var data []testData
+
+	// create test configs with default settings
+	for i := 0; i < 18; i++ {
+		data = append(data, testData{newQemuConfig(), false, true})
+	}
+
+	// customize configs
+	data[1].config.SeccompSandbox = "enable=1"
+	data[2].config.InitrdPath = ""
+	data[2].config.ImagePath = testQemuImagePath
+	data[3].config.Debug = true
+	data[4].config.IOMMU = true
+	data[5].config.GuestMemoryDumpPath = "/tmp"
+	data[6].config.DisableGuestSeLinux = false
+	data[6].configMatch = false
+	data[7].config.PCIeRootPort = 1
+	data[8].config.EnableVhostUserStore = true
+	data[8].config.HugePages = true
+	data[9].config.EnableVhostUserStore = true
+	data[9].config.HugePages = false
+	data[9].expectError = true
+	data[10].config.BootToBeTemplate = true
+	data[11].config.BootFromTemplate = true
+	data[12].config.BootToBeTemplate = true
+	data[12].config.SharedFS = config.VirtioFS
+	data[12].expectError = true
+	data[13].config.FileBackedMemRootDir = "/tmp/xyzabc"
+	data[13].config.HugePages = true
+	data[14].config.SharedFS = config.VirtioFS
+	data[15].config.BlockDeviceDriver = ""
+	data[16].config.SharedFS = config.VirtioFSNydus
+	data[17].config.VMid = "testSandbox"
```

Comment:
This is admittedly very compact, but I find it very difficult to read.

Please could you change this to match more like what is shown in https://github.com/kata-containers/kata-containers/blob/main/docs/Unit-Test-Advice.md#golang-table-driven-tests by using partial struct initialisation:

```go
data := []testData{
    {},
    {config: HypervisorConfig{SeccompSandbox: "enable=1"}},
    {config: HypervisorConfig{ImagePath: testQemuImagePath}},

    // ...
}
```

This will also allow you to remove the `for` loop with the hard-coded upper index bound above, so the code will be safer and easier to maintain.

Reply:
This is all fine if all you are setting is a single config entry. It becomes ugly once you start adding more config options:

```
    data := []testData{
        { false, true, newQemuConfig() },
        { false, false, HypervisorConfig{ SeccompSandbox = "enable=1" } },
        { false, false, HypervisorConfig {  InitrdPath = "",
                                            ImagePath = testQemuImagePath } },
        { false, false, HypervisorConfig { Debug = true },

    }
```

I find this format to be a lot harder to read then default-initializing all test-configs and then just changing options by index. With partial init only you are also losing the check `assert.Exactly(d.config, q.config, msg)` because it fails for every one except the default one generated by `newQemuConfig()`. Furthermore, without default settings being present in every config the code coverage is lower because the code bails earlier, so you end up having to define the same settings over and over and over for every config further making it harder to understand which option is actually being tested.

Reply:
True, but when we start adding large chunks we tend to create a few variables to make this clearer. For example see how we use `longPath`, `longestPath` and `pathTooLong` here:

- https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/utils/utils_test.go#L276..L292

Reply:
Sure, I can do that, IMHO this approach adds more lines than necessary but if you think it makes it more readable I won't argue